### ISSUE TITLE
“Different F1 results (0.72 vs 0.77-0.78) for the same model and the same evaluation script when switching from pip-installed ultralytics 8.3.190 to local git-cloned ultralytics 8.3.199”

### ISSUE THREAD

**Issue by YIYIYI14736** [ultralytics/ultralytics#22200]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi team,

I’m seeing a reproducible gap in F1 score on my **own test set** when I only change the way ultralytics is installed—**nothing else** (same `.pt`, same images, same labels, same evaluation code, same hardware, same CUDA/cuDNN, same Python env except for the ultralytics package).

| Installation method | ultralytics version | mF1 on my test set |
|---------------------|---------------------|---------------------|
| `pip install ultralytics==8.3.190` | 8.3.190 | **0.72** |
| `git clone` repo (tag `v8.3.199`) and add repo root to `PYTHONPATH` | 8.3.199 | **0.77 – 0.78** |

The evaluation script is **identical** (copied above); it simply calls  
`model = YOLO("best.pt")`  
`results = model.predict(...)`  
and then runs my own matching logic.

I already double-checked:
- `args.conf`, `args.nms_iou`, `imgsz`, `device` are identical  
- `best.pt` is byte-for-byte the same file  
- random seeds are fixed (`torch.manual_seed(42); np.random.seed(42)`)  
- single GPU, deterministic cuDNN, `torch.use_deterministic_algorithms(True)`  
- no AMP vs FP16 difference (tried `--half` and without; gap remains)  
- diffing the two installed packages shows changes in `nn/modules/head.py`, `engine/predictor.py`, `data/augment.py`, etc. between 8.3.190 and 8.3.199.

Could you help me understand **which commit or code change between 8.3.190 and 8.3.199** could explain the ~0.05 F1 jump?  
Is there any silent change in NMS, letterbox padding, or confidence rounding that might shift the boxes that survive into my evaluation?

Thanks!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @YIYIYI14736, thanks for reporting this and for the detailed investigation 🚀 This is an automated response to help move things forward quickly—an Ultralytics engineer will also assist you here soon.

Since this appears to be a 🐛 Bug Report with version-dependent behavior, could you please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) so we can reproduce the F1 gap on our side?

To expedite triage, please include:
- A minimal script showing how you load and run inference (Python or CLI), plus exact command-line args used
- The exact versions and install methods you tested (e.g., `pip install ultralytics==8.3.190` vs local `v8.3.199`)
- The output of `pip freeze` and `python -c "import ultralytics, torch; print(ultralytics.__version__, torch.__version__)"` and GPU info
- The output of `from ultralytics.utils.checks import check_requirements; check_requirements()` or `yolo checks`
- A tiny, anonymized subset of your test images and labels (just enough to reproduce the delta)
- The specific inference settings you rely on (e.g., `conf`, `iou`, `imgsz`, `device`, `half`, `max_det`, NMS class-agnostic or not)
- Whether Python vs CLI predicts show the same difference using identical settings

In the meantime, for general usage and verification:
- See the [Docs](https://docs.ultralytics.com/) for [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples that you can use as a baseline to compare runs across versions.
- If this turns out to be training/evaluation setup related, double-check your pipeline against our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community for more support:
- Real-time chat on [Discord](https://discord.com/invite/ultralytics) 🎧
- In-depth Q&A on [Discourse](https://community.ultralytics.com/)
- Community threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify the issue persists in the newest release:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- Amazon Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- Docker Image. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
> `git clone` repo (tag `v8.3.199`) and add repo root to `PYTHONPATH`

This is not how you use the cloned repo. You should install it with `pip`.
https://docs.ultralytics.com/quickstart/#git-clone
